File: library/src/main/java/moe/leekcake/live2dforjvm/core/type/AutoPointer.java
package moe.leekcake.live2dforjvm.core.type;

import moe.leekcake.live2dforjvm.core.jni.MemoryAccessJNI;

public abstract class AutoPointer {
    public final long pointer;
    private boolean isReleased = false;
    public AutoPointer(long pointer) {
        this.pointer = pointer;
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        release();
    }

    private void release() {
        if(isReleased) return;
        isReleased = true;
        releasePointer();
    }

    protected void releasePointer() {
        MemoryAccessJNI.deAllocate(pointer);
    }
}

File: library/src/main/kotlin/moe/leekcake/live2dforjvm/type/AutoPointer.kt
package moe.leekcake.live2dforjvm.type

import moe.leekcake.live2dforjvm.MemoryAccessJNI

/**
 * try to auto-release c pointer with java 'finalize' method,
 */
abstract class AutoPointer(val pointer: Long) {
    private var released: Boolean = false

    //Java will call it, may be?
    @Suppress("Unused")
    protected fun finalize() {
        release()
    }

    open fun release() {
        if (released) return
        MemoryAccessJNI.deAllocate(pointer)
        released = true
    }
}